pkg/controller/atlasprivateendpoint/atlasprivateendpoint_controller.go

```diff
@@ -0,0 +1,343 @@
+		return r.skip(ctx, akoPrivateEndpoint)
+	}
+
+	conditions := api.InitCondition(akoPrivateEndpoint, api.FalseCondition(api.ReadyType))
+	workflowCtx := workflow.NewContext(r.Log, conditions, ctx)
+	defer statushandler.Update(workflowCtx, r.Client, r.EventRecorder, akoPrivateEndpoint)
+
+	isValid := customresource.ValidateResourceVersion(workflowCtx, akoPrivateEndpoint, r.Log)
+	if !isValid.IsOk() {
+		return r.invalidate(isValid)
+	}
+
+	if !r.AtlasProvider.IsResourceSupported(akoPrivateEndpoint) {
+		return r.unsupport(workflowCtx)
+	}
+
+	var atlasProject *project.Project
+	var err error
+	if akoPrivateEndpoint.Spec.ExternalProject != nil {
+		atlasProject, err = r.getProjectFromAtlas(ctx, akoPrivateEndpoint)
+	} else {
+		atlasProject, err = r.getProjectFromKube(ctx, akoPrivateEndpoint)
+	}
+	if err != nil {
+		return r.terminate(workflowCtx, akoPrivateEndpoint, nil, api.ReadyType, workflow.AtlasAPIAccessNotConfigured, err)
+	}
+
+	return r.handlePrivateEndpointService(workflowCtx, atlasProject.ID, akoPrivateEndpoint)
+}
+
+func (r *AtlasPrivateEndpointReconciler) getProjectFromAtlas(ctx context.Context, akoPrivateEndpoint *akov2.AtlasPrivateEndpoint) (*project.Project, error) {
+	sdkClient, _, err := r.AtlasProvider.SdkClient(
+		ctx,
+		&client.ObjectKey{Namespace: akoPrivateEndpoint.Namespace, Name: akoPrivateEndpoint.Credentials().Name},
+		r.Log,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	projectService := project.NewProjectAPIService(sdkClient.ProjectsApi)
+	r.privateEndpointService = privateendpoint.NewPrivateEndpointAPI(sdkClient.PrivateEndpointServicesApi)
+
+	atlasProject, err := projectService.GetProject(ctx, akoPrivateEndpoint.Spec.ExternalProject.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	return atlasProject, nil
+}
+
+func (r *AtlasPrivateEndpointReconciler) getProjectFromKube(ctx context.Context, akoPrivateEndpoint *akov2.AtlasPrivateEndpoint) (*project.Project, error) {
+	atlasProject := &akov2.AtlasProject{}
+	if err := r.Client.Get(ctx, akoPrivateEndpoint.AtlasProjectObjectKey(), atlasProject); err != nil {
+		return nil, err
+	}
+
+	credentialsSecret, err := customresource.ComputeSecret(atlasProject, akoPrivateEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	sdkClient, orgID, err := r.AtlasProvider.SdkClient(ctx, credentialsSecret, r.Log)
+	if err != nil {
+		return nil, err
+	}
+
+	r.privateEndpointService = privateendpoint.NewPrivateEndpointAPI(sdkClient.PrivateEndpointServicesApi)
+
+	return project.NewProject(atlasProject, orgID), nil
+}
+
+func (r *AtlasPrivateEndpointReconciler) skip(ctx context.Context, akoPrivateEndpoint *akov2.AtlasPrivateEndpoint) (ctrl.Result, error) {
+	r.Log.Infow(fmt.Sprintf("-> Skipping AtlasPrivateEndpoint reconciliation as annotation %s=%s", customresource.ReconciliationPolicyAnnotation, customresource.ReconciliationPolicySkip), "spec", akoPrivateEndpoint.Spec)
+	if !akoPrivateEndpoint.GetDeletionTimestamp().IsZero() {
```

Comment:
`skip` is a transition method while this code is asserting whether another transition is supposed to be executed. `skip` should only contain setting skip-state, not transition to other state. Checking deletion timestamp has to be done outside this method.